libs/vkd3d/bundle.c

```diff
@@ -0,0 +1,1754 @@
+/*
+ * Copyright 2021 Philip Rebohle for Valve Corporation
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+#define VKD3D_DBG_CHANNEL VKD3D_DBG_CHANNEL_API
+
+#include "vkd3d_private.h"
+
+/* ID3D12CommandAllocator */
+static inline struct d3d12_bundle_allocator *impl_from_ID3D12CommandAllocator(ID3D12CommandAllocator *iface)
+{
+    return CONTAINING_RECORD(iface, struct d3d12_bundle_allocator, ID3D12CommandAllocator_iface);
+}
+
+static void *d3d12_bundle_allocator_alloc_chunk_data(struct d3d12_bundle_allocator *allocator, size_t size)
+{
+    size_t chunk_offset = 0;
+    void *chunk = NULL;
+
+    size = align(size, VKD3D_BUNDLE_COMMAND_ALIGNMENT);
+
+    if (allocator->chunks_count)
+    {
+        chunk = allocator->chunks[allocator->chunks_count - 1];
+        chunk_offset = allocator->chunk_offset;
+    }
+
+    if (!chunk || chunk_offset + size > VKD3D_BUNDLE_CHUNK_SIZE)
+    {
+        if (!vkd3d_array_reserve((void **)&allocator->chunks, &allocator->chunks_size,
+                allocator->chunks_count + 1, sizeof(*allocator->chunks)))
+            return NULL;
+
+        if (!(chunk = vkd3d_malloc(VKD3D_BUNDLE_CHUNK_SIZE)))
+            return NULL;
+
+        allocator->chunks[allocator->chunks_count++] = chunk;
+        allocator->chunk_offset = chunk_offset = 0;
+    }
+
+    allocator->chunk_offset = chunk_offset + size;
+    return void_ptr_offset(chunk, chunk_offset);
+}
+
+static void d3d12_bundle_allocator_free_chunks(struct d3d12_bundle_allocator *allocator)
+{
+    size_t i;
+
+    for (i = 0; i < allocator->chunks_count; i++)
+        vkd3d_free(allocator->chunks[i]);
+
+    allocator->chunks_count = 0;
```

Comment:
Shouldn't allocator->chunks also be freed?

Reply:
The original idea was to recycle the array but right now we're not freeing it anywhere, so yeah, added that as well.